fw/preamp/src/audio.c

```diff
@@ -48,104 +45,120 @@ const I2CReg zone_right_[NUM_ZONES] = {
 }
 
-// Writes volume level to the volume ICs via the internal I2C bus
-void writeVolume(size_t zone, uint8_t vol) {
-  // We can't write to the volume registers if they are disabled
-  if (!inStandby()) {
-    writeI2C2(zone_left_[zone], vol);
-    writeI2C2(zone_right_[zone], vol);
+uint8_t getZoneVolume(size_t zone) {
+  if (muted(zone)) {
+    // If muted the real volume will be VOL_MUTE.
+    // Instead, report to the user the value which will be set upon unmuting.
+    return vol_req_[zone];
+  } else {
+    return vol_[zone];
   }
 }
 
-// Standby all amps at once
-void standby(bool standby) {
-  bool prev_stby_state = inStandby();
+// All amps must be in standby together due to the SYNCLK
+static void standby(bool standby) {
   for (size_t zone = 0; zone < NUM_ZONES; zone++) {
     // Set pin low to standby
     writePin(zone_standby_[zone], !standby);
   }
-  // After returning from standby we need to configure each of the volumes again
-  if (prev_stby_state && !standby) {
-    for (size_t zone = 0; zone < NUM_ZONES; zone++) {
-      writeVolume(zone, volumes[zone]);
-    }
-  }
 }
 
 // Checks if any of the zones are in standby
 bool inStandby() {
-  bool in_stby = false;
   for (size_t zone = 0; zone < NUM_ZONES; zone++) {
-    in_stby = in_stby || (!readPin(zone_standby_[zone]));
+    // Standby pins are active-low
+    if (!readPin(zone_standby_[zone])) {
+      return true;
+    }
   }
-  return in_stby;
-  // TODO: Shortcut
+  return false;
 }
 
-// Initialize each zone's volume state (does not write to volume control ICs)
-void initZones() {
-  for (size_t zone = 0; zone < NUM_ZONES; zone++) {
-    volumes[zone] = DEFAULT_VOL;
-    mute(zone, true);
-    setZoneSource(zone, DEFAULT_SOURCE);
+// Returns true if any zone is unmuted
+static bool anyZoneOn() {
+  for (uint8_t zone = 0; zone < NUM_ZONES; zone++) {
+    if (!muted(zone)) {
+      return true;
+    }
   }
-  standby(true);
+  return false;
 }
 
-// Set a zone's volume (required I2C write)
-void setZoneVolume(size_t zone, uint8_t vol) {
-#ifdef AUTO_MUTE_CTRL
-  // Automatic On-Off control
-  if (vol > 78) {
-    mute(zone, true);
-  } else if (!IsOn(zone)) {
-    mute(zone, false);
-  }
-#endif
-
-  // Keep track of the volume so it is not lost when we standby
-  volumes[zone] = vol;
+// Mute the specified zone
+void mute(size_t zone, bool mute) {
+  // Set pin low to mute
+  writePin(zone_mute_[zone], !mute);
 
-  // Actually write the volume to the volume control IC
-  writeVolume(zone, vol);
+  // If no zones are on, standby
+  standby(!anyZoneOn());
 }
 
-uint8_t getZoneVolume(size_t zone) {
-  return volumes[zone];
+bool muted(size_t zone) {
+  // Mute pins are active-low
+  return !readPin(zone_mute_[zone]);
 }
 
 // Connect a Zone to a Source
 void setZoneSource(size_t zone, size_t src) {
   // Mute the zone during the switch to avoid an audible pop
-  bool was_muted = !isOn(zone);
+  bool was_muted = muted(zone);
   mute(zone, !was_muted);
```

Comment:
This code doesn't do what the comment says, try this
```suggestion
  mute(zone, true);
```